File: tests/test_parser.py
from unittest import TestCase
from vsi.parser import Parser
from vsi.lexer import Token
from vsi.api import lex
from vsi.ast import (
    PrintStatement,
    AST,
    Integer,
    Varexpr,
    BinOpexpr,
    RelOpexpr,
    Andexpr,
    Orexpr,
    Notexpr,
    AssignStatement,
    IfStatement,
    WhileStatement,
    Float
)


class TestParser(TestCase):

    def test___repr___method(self):
        tokens = lex("1 + 1;")
        ast = AST()
        parser = Parser(tokens)
        parser_repr = repr(parser)
        expected_repr = "Parser({}, 0, {})".format(tokens, ast)
        self.assertEqual(parser_repr, expected_repr)

    def test_parse_method(self):
        tokens = lex("print 1 + 1;")
        parser = Parser(tokens)
        ast = parser.parse()
        expected_ast = AST([PrintStatement(
                            BinOpexpr("+",
                                      Integer(1),
                                      Integer(1)
                                      )
                            )
                            ])
        self.assertEqual(ast, expected_ast)

    def test_parse_statements_method(self):
        tokens = lex("print 1;")
        parser = Parser(tokens)
        parser.parse()
        expected_ast = AST([PrintStatement(Integer(1))])
        self.assertEqual(parser.ast, expected_ast)

    def test_found_method(self):
        tokens = lex("x := 1 + 1;")
        parser = Parser(tokens)
        self.assertTrue(parser.found("ID", by="type"))
        parser.advance(by=2)
        self.assertTrue(parser.found(1, by="value"))

    def test_found_one_of_method(self):
        tokens = lex("x := 1 + 1;")
        parser = Parser(tokens)
        self.assertTrue(parser.found_one_of(["ID", "RESERVED"], by="type"))
        parser.advance(by=2)
        self.assertTrue(parser.found_one_of(["NUMBER", "OPERATOR"], by="type"))

    def test_expect_method(self):
        tokens = lex("print 1;")
        parser = Parser(tokens)
        self.assertTrue(parser.expect("RESERVED", by="type"))

    def test_expect_one_of_method(self):
        tokens = lex("print 1;")
        parser = Parser(tokens)
        self.assertTrue(parser.expect_one_of(["RESERVED", "ID"], by="type"))

    def test_optional_method(self):
        tokens = lex("if x > 10 then\nx:=1\nelse x:=10")
        parser = Parser(tokens)
        self.assertIsNone(parser.optional("else"))
        parser.advance(by=8)
        self.assertTrue(parser.optional("else"))

    def test_advance_method(self):
        tokens = lex("x := 1")
        parser = Parser(tokens)
        parser.advance()
        self.assertEqual(parser.pos, 1)

    def test_token_ahead_method(self):
        tokens = lex("x := 1")
        parser = Parser(tokens)
        parser.advance()
        self.assertEqual(parser.token_ahead().value, 1)

    def test_parse_if_statement_method(self):
        tokens = lex("if x > 22 then\nprint x;\ndone\n")
        parser = Parser(tokens)
        parser.advance()
        ast = parser.parse_if_statement()
        expected_ast = IfStatement(
            RelOpexpr(">", Varexpr("x"), Integer(22)),
            AST([PrintStatement(Varexpr("x"))]),
            false_stmt=None
        )
        self.assertEqual(ast, expected_ast)

    def parse_if_statement_method_with_else_statement(self):
        tokens = lex("if x > 22 then\nprint x;\nelse\nprint 0;\ndone\n")
        parser = Parser(tokens)
        parser.advance()
        ast = parser.parse_if_statement()
        expected_ast = IfStatement(
            RelOpexpr(">", Varexpr("x"), Integer(22)),
            AST([PrintStatement(Varexpr("x"))]),
            AST([PrintStatement(Integer(0))])
        )
        self.assertEqual(ast, expected_ast)

    def test_parse_while_statement_method(self):
        tokens = lex("while x > 22 do\nprint x;\nx := x - 1;\ndone\n")
        parser = Parser(tokens)
        parser.advance()
        ast = parser.parse_while_statement()
        expected_ast = WhileStatement(
            RelOpexpr(">", Varexpr("x"), Integer(22)),
            AST([PrintStatement(Varexpr("x")),
                 AssignStatement("x", BinOpexpr("-", Varexpr("x"), Integer(1)))
                 ]),
        )
        self.assertEqual(ast, expected_ast)

    def test_parse_var_assignment_method(self):
        tokens = lex("x := 10;")
        parser = Parser(tokens)
        ast = parser.parse_var_assignment()
        expected_ast = AssignStatement("x", Integer(10))
        self.assertEqual(ast, expected_ast)

    def test_parse_print_statement_method(self):
        tokens = lex("print 10;")
        parser = Parser(tokens)
        parser.advance()
        ast = parser.parse_print_statement()
        expected_ast = PrintStatement(Integer(10))
        self.assertEqual(ast, expected_ast)

    def test_parse_expr_method(self):
        tokens = lex("1+1;")
        parser = Parser(tokens)
        ast = parser.parse_expr()
        expected_ast = BinOpexpr("+", Integer(1), Integer(1))
        self.assertEqual(ast, expected_ast)

    def test_create_ast_from_postfix_method(self):
        postfix = [Token(1, "NUMBER", "INT"),
                   Token(1, "NUMBER", "INT"),
                   Token("+", "RESERVED", "OPERATOR")
                   ]
        parser = Parser([])
        ast = parser.create_ast_from_postfix(postfix)
        expected_ast = BinOpexpr("+", Integer(1), Integer(1))
        self.assertEqual(ast, expected_ast)

    def test_create_ast_from_expr_method(self):
        parser = Parser([])
        ast = parser.create_ast_from_expr("+", Integer(1), Integer(1))
        expected_ast = BinOpexpr("+", Integer(1), Integer(1))
        self.assertEqual(ast, expected_ast)

File: Vsi.Tests/TestParser.cs
using System.Collections.Generic;
using Xunit;
using Vsi.Core;

namespace Vsi.Tests
{
    public class TestParser
    {
        [Fact]
        public void TestParserReprMethod()
        {
            var tokens = Api.Lex("1 + 1;");
            var parser = new Parser(tokens);
            var repr = parser.ToString();
            Assert.Contains("Parser([", repr);
        }

        [Fact]
        public void TestParseMethod()
        {
            var tokens = Api.Lex("print 1 + 1;");
            var parser = new Parser(tokens);
            var ast = parser.Parse();
            var expectedAst = new Ast(new List<AstNode>
            {
                new PrintStatement(new BinOpExpr("+", new Integer(1), new Integer(1)))
            });
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestParseStatementsMethod()
        {
            var tokens = Api.Lex("print 1;");
            var parser = new Parser(tokens);
            parser.Parse();
            var expectedAst = new Ast(new List<AstNode> { new PrintStatement(new Integer(1)) });
            Assert.Equal(expectedAst, parser.Ast);
        }

        [Fact]
        public void TestFoundMethod()
        {
            var tokens = Api.Lex("x := 1 + 1;");
            var parser = new Parser(tokens);
            Assert.True(parser.Found(TokenType.Id));
            parser.Advance(2);
            Assert.True(parser.Found("1"));
        }

        [Fact]
        public void TestFoundOneOfMethod()
        {
            var tokens = Api.Lex("x := 1 + 1;");
            var parser = new Parser(tokens);
            Assert.True(parser.FoundOneOf(new[] { TokenType.Id, TokenType.Reserved }));
            parser.Advance(1); // at :=
            Assert.True(parser.FoundOneOf(new[] { TokenSpecificType.Operator, TokenSpecificType.Symbol }));
        }

        [Fact]
        public void TestExpectMethod()
        {
            var tokens = Api.Lex("print 1;");
            var parser = new Parser(tokens);
            Assert.True(parser.Expect(TokenType.Reserved));
        }

        [Fact]
        public void TestExpectOneOfMethod()
        {
            var tokens = Api.Lex("print 1;");
            var parser = new Parser(tokens);
            Assert.True(parser.ExpectOneOf(new[] { TokenType.Reserved, TokenType.Id }));
        }

        [Fact]
        public void TestOptionalMethod()
        {
            var tokens = Api.Lex("if x > 10 then\nx:=1\nelse x:=10");
            var parser = new Parser(tokens);
            Assert.False(parser.Optional("else"));
            parser.Advance(8);
            Assert.True(parser.Optional("else"));
        }

        [Fact]
        public void TestAdvanceMethod()
        {
            var tokens = Api.Lex("x := 1");
            var parser = new Parser(tokens);
            parser.Advance();
            Assert.Equal(1, parser.Pos);
        }

        [Fact]
        public void TestTokenAheadMethod()
        {
            var tokens = Api.Lex("x := 1");
            var parser = new Parser(tokens);
            parser.Advance();
            Assert.Equal(1, parser.TokenAhead()?.Value);
        }

        [Fact]
        public void TestParseIfStatementMethod()
        {
            var tokens = Api.Lex("if x > 22 then\nprint x;\ndone\n");
            var parser = new Parser(tokens);
            parser.Advance();
            var ast = parser.ParseIfStatement();
            var expectedAst = new IfStatement(
                new RelOpExpr(">", new VarExpr("x"), new Integer(22)),
                new Ast(new List<AstNode> { new PrintStatement(new VarExpr("x")) }),
                null
            );
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void ParseIfStatementMethodWithElseStatement()
        {
            var tokens = Api.Lex("if x > 22 then\nprint x;\nelse\nprint 0;\ndone\n");
            var parser = new Parser(tokens);
            parser.Advance();
            var ast = parser.ParseIfStatement();
            var expectedAst = new IfStatement(
                new RelOpExpr(">", new VarExpr("x"), new Integer(22)),
                new Ast(new List<AstNode> { new PrintStatement(new VarExpr("x")) }),
                new Ast(new List<AstNode> { new PrintStatement(new Integer(0)) })
            );
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestParseWhileStatementMethod()
        {
            var tokens = Api.Lex("while x > 22 do\nprint x;\nx := x - 1;\ndone\n");
            var parser = new Parser(tokens);
            parser.Advance();
            var ast = parser.ParseWhileStatement();
            var expectedAst = new WhileStatement(
                new RelOpExpr(">", new VarExpr("x"), new Integer(22)),
                new Ast(new List<AstNode>
                {
                    new PrintStatement(new VarExpr("x")),
                    new AssignStatement("x", new BinOpExpr("-", new VarExpr("x"), new Integer(1)))
                })
            );
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestParseVarAssignmentMethod()
        {
            var tokens = Api.Lex("x := 10;");
            var parser = new Parser(tokens);
            var ast = parser.ParseVarAssignment();
            var expectedAst = new AssignStatement("x", new Integer(10));
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestParsePrintStatementMethod()
        {
            var tokens = Api.Lex("print 10;");
            var parser = new Parser(tokens);
            parser.Advance();
            var ast = parser.ParsePrintStatement();
            var expectedAst = new PrintStatement(new Integer(10));
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestParseExprMethod()
        {
            var tokens = Api.Lex("1+1;");
            var parser = new Parser(tokens);
            var ast = parser.ParseExpr();
            var expectedAst = new BinOpExpr("+", new Integer(1), new Integer(1));
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestCreateAstFromPostfixMethod()
        {
            var postfix = new List<Token>
            {
                new Token(1, TokenType.Number, TokenSpecificType.Int),
                new Token(1, TokenType.Number, TokenSpecificType.Int),
                new Token("+", TokenType.Reserved, TokenSpecificType.Operator)
            };
            var parser = new Parser(new List<Token>());
            var ast = parser.CreateAstFromPostfix(postfix);
            var expectedAst = new BinOpExpr("+", new Integer(1), new Integer(1));
            Assert.Equal(expectedAst, ast);
        }

        [Fact]
        public void TestCreateAstFromExprMethod()
        {
            var parser = new Parser(new List<Token>());
            var ast = parser.CreateAstFromExpr("+", new Integer(1), new Integer(1));
            var expectedAst = new BinOpExpr("+", new Integer(1), new Integer(1));
            Assert.Equal(expectedAst, ast);
        }
    }
}
